daemon/graphdriver/devmapper/deviceset.go

```diff
@@ -1818,10 +1818,59 @@ func (devices *DeviceSet) AddDevice(hash, baseHash string) error {
 		return fmt.Errorf("devmapper: device %s already exists. Deleted=%v", hash, info.Deleted)
 	}
 
-	if err := devices.createRegisterSnapDevice(hash, baseInfo); err != nil {
+	devinfo := &devInfo{}
+
+	if err := devices.parseStorageOpt(storageOpt, devinfo); err != nil {
+		return err
+	}
+
+	if devinfo.Size == 0 {
+		devinfo.Size = baseInfo.Size
+	}
+
+	if devinfo.Size < baseInfo.Size {
+		return fmt.Errorf("devmapper: Container size cannot be smaller than %s", units.HumanSize(float64(baseInfo.Size)))
+	}
+
+	if err := devices.createRegisterSnapDevice(hash, baseInfo, devinfo.Size); err != nil {
 		return err
 	}
 
+	// Grow the container rootfs.
+	if devinfo.Size > 0 {
+		info, err := devices.lookupDevice(hash)
+		if info == nil {
+			return fmt.Errorf("Failed to lookup device %s:%v", hash, err)
+		}
+
+		if err := devices.growFS(info); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (devices *DeviceSet) parseStorageOpt(storageOpt []string, devinfo *devInfo) error {
+
+	// Read size to change the block device size per container.
+	var size int64
+	var err error
+	for _, option := range storageOpt {
+		val := strings.SplitN(option, ":", 2)
```

Comment:
can you split w/o the 2? and check len(val) != 2 and error out here? I think you could get something like val[0] = "size" and val[1] = "1G:test" and fails below (didn't check :))